File: services/typescript/src/service/workspace-service.ts
import { InvalidInputError } from "../repository/errors.js";
import {
  WorkspaceRepo,
  type CreateWorkspaceParams,
  type ListWorkspacesParams,
  type UpdateWorkspaceParams,
  type Workspace,
  type WorkspaceList,
} from "../repository/workspace-repo.js";
import type { Logger } from "pino";

export class WorkspaceService {
  constructor(
    private repo: WorkspaceRepo,
    private logger: Logger,
  ) {}

  async create(params: CreateWorkspaceParams): Promise<Workspace> {
    if (!params.name) {
      throw new InvalidInputError("invalid input: name is required");
    }
    if (!params.slug) {
      throw new InvalidInputError("invalid input: slug is required");
    }
    this.logger.debug({ name: params.name, slug: params.slug }, "creating workspace");
    return this.repo.create(params);
  }

  async getById(id: string): Promise<Workspace> {
    return this.repo.getById(id);
  }

  async list(params: ListWorkspacesParams): Promise<WorkspaceList> {
    return this.repo.list(params);
  }

  async update(params: UpdateWorkspaceParams): Promise<Workspace> {
    if (!params.name) {
      throw new InvalidInputError("invalid input: name is required");
    }
    return this.repo.update(params);
  }

  async delete(id: string): Promise<void> {
    return this.repo.delete(id);
  }
}

File: services/kotlin/src/main/kotlin/com/corestack/service/WorkspaceService.kt
package com.corestack.service

import com.corestack.repository.*
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.UUID

private val logger = KotlinLogging.logger {}

class WorkspaceService(private val repo: WorkspaceRepo) {

    suspend fun create(params: CreateWorkspaceParams): Workspace {
        if (params.name.isBlank()) throw InvalidInputError("invalid input: name is required")
        if (params.slug.isBlank()) throw InvalidInputError("invalid input: slug is required")
        logger.debug { "creating workspace name=${params.name} slug=${params.slug}" }
        return repo.create(params)
    }

    suspend fun getById(id: UUID): Workspace = repo.getById(id)

    suspend fun list(params: ListWorkspacesParams): WorkspaceList = repo.list(params)

    suspend fun update(params: UpdateWorkspaceParams): Workspace {
        if (params.name.isBlank()) throw InvalidInputError("invalid input: name is required")
        return repo.update(params)
    }

    suspend fun delete(id: UUID) = repo.delete(id)
}
